golang/src/user-sessions/main.go

```diff
@@ -0,0 +1,393 @@
+}
+
+func (u *UserSessionsApp) getUserSessionForRequest(username string, wrap bool) ([]byte, error) {
+	sessions, err := u.getSessions(username)
+	if err != nil {
+		return nil, fmt.Errorf("Error getting sessions for username %s: %s", username, err)
+	}
+	var retval UserSessionRecord
+	if len(sessions) >= 1 {
+		retval = sessions[0]
+	}
+	response, err := convert(&retval, wrap)
+	if err != nil {
+		return nil, fmt.Errorf("Error generating response for username %s: %s", username, err)
+	}
+	var jsoned []byte
+	if len(response) > 0 {
+		jsoned, err = json.Marshal(response)
+		if err != nil {
+			return nil, fmt.Errorf("Error generating session JSON for user %s: %s", username, err)
+		}
+	} else {
+		jsoned = []byte("{}")
+	}
+	return jsoned, nil
+}
+
+// GetRequest handles writing out a user's session as a response.
+func (u *UserSessionsApp) GetRequest(writer http.ResponseWriter, r *http.Request) {
+	var (
+		username   string
+		userExists bool
+		err        error
+		ok         bool
+		v          = mux.Vars(r)
+	)
+	if username, ok = v["username"]; !ok {
+		badRequest(writer, "Missing username in URL")
+		return
+	}
+	logcabin.Info.Printf("Getting user session for %s", username)
+	if userExists, err = queries.IsUser(u.db, username); err != nil {
+		badRequest(writer, fmt.Sprintf("Error checking for username %s: %s", username, err))
+		return
+	}
+	if !userExists {
+		badRequest(writer, fmt.Sprintf("User %s does not exist", username))
+		return
+	}
+	jsoned, err := u.getUserSessionForRequest(username, false)
+	if err != nil {
+		errored(writer, err.Error())
+	}
+	writer.Write(jsoned)
+}
+
+// PutRequest handles creating a new user session.
+func (u *UserSessionsApp) PutRequest(writer http.ResponseWriter, r *http.Request) {
+	u.PostRequest(writer, r)
+}
+
+// PostRequest handles modifying an existing user session.
+func (u *UserSessionsApp) PostRequest(writer http.ResponseWriter, r *http.Request) {
+	var (
+		username   string
+		userExists bool
+		hasSession bool
+		err        error
+		ok         bool
+		v          = mux.Vars(r)
+	)
+	if username, ok = v["username"]; !ok {
+		badRequest(writer, "Missing username in URL")
+		return
+	}
+	if userExists, err = queries.IsUser(u.db, username); err != nil {
+		badRequest(writer, fmt.Sprintf("Error checking for username %s: %s", username, err))
+		return
+	}
+	if !userExists {
+		badRequest(writer, fmt.Sprintf("User %s does not exist", username))
+		return
+	}
+	if hasSession, err = u.hasSessions(username); err != nil {
+		errored(writer, fmt.Sprintf("Error checking session for user %s: %s", username, err))
+		return
+	}
+	var checked map[string]interface{}
+	bodyBuffer, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		errored(writer, fmt.Sprintf("Error reading body: %s", err))
+		return
+	}
+	if err = json.Unmarshal(bodyBuffer, &checked); err != nil {
+		errored(writer, fmt.Sprintf("Error parsing request body: %s", err))
+		return
+	}
+	bodyString := string(bodyBuffer)
+	if !hasSession {
```

Comment:
With us depending on this value for an upsert-type operation, it would be best if this was in a transaction, I think. Not sure how much of a pain that would be to handle.